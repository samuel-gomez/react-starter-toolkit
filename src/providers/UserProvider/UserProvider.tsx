import { createContext, ReactNode, useMemo } from 'react';
import { useOidcUser } from '@axa-fr/react-oidc/dist/User';
import isEmpty from 'lodash/isEmpty';
import { PROFILS } from 'shared/constants';

export type TUserContext = ReturnType<typeof extractDataFromOAuthToken> & {
  isEnabled?: boolean;
  isLoading?: boolean;
};

export const UserContext = createContext<TUserContext>({
  authName: '',
  authRole: '',
  authUid: '',
  isEnabled: true,
  isLoading: true,
});
UserContext.displayName = 'UserContext';

const NON_CONNECTE = 'Non ConnectÃ©';

type ToidcUser = {
  member_of?: string[];
  name?: string;
  axa_uid_racf?: string;
};

/**
 * getAuthName
 */

type TgetAuthName = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthName = ({ oidcUser, isEmptyFn = isEmpty }: TgetAuthName) => (!isEmptyFn(oidcUser?.name ?? '') ? oidcUser.name : NON_CONNECTE);

/**
 * setAuthRole
 */
type TsetAuthRole = {
  memberOf: string;
  profils?: typeof PROFILS;
};

export const setAuthRole = ({ memberOf, profils = PROFILS }: TsetAuthRole) =>
  profils.map(profil => (memberOf.search(`${profil}`) !== -1 ? profil : '')).join('');

/**
 * getAuthRole
 */

type TgetAuthRole = {
  oidcUser: ToidcUser;
  setAuthRoleFn?: typeof setAuthRole;
};

export const getAuthRole = ({ oidcUser, setAuthRoleFn = setAuthRole }: TgetAuthRole) => setAuthRoleFn({ memberOf: oidcUser?.member_of?.[0] ?? '' });

/**
 * getAuthUid
 */

type TgetAuthUid = {
  oidcUser: ToidcUser;
  isEmptyFn?: typeof isEmpty;
};

export const getAuthUid = ({ oidcUser, isEmptyFn = isEmpty }: TgetAuthUid) =>
  (!isEmptyFn(oidcUser?.axa_uid_racf ?? '') ? oidcUser.axa_uid_racf : '') || '';

/**
 * extractDataFromOAuthToken
 * @param {Object} oidcUser
 */

type TextractDataFromOAuthToken = {
  oidcUser: ToidcUser;
  getAuthNameFn?: typeof getAuthName;
  getAuthRoleFn?: typeof getAuthRole;
  getAuthUidFn?: typeof getAuthUid;
};

const extractDataFromOAuthToken = ({
  oidcUser,
  getAuthNameFn = getAuthName,
  getAuthRoleFn = getAuthRole,
  getAuthUidFn = getAuthUid,
}: TextractDataFromOAuthToken) => ({
  authName: getAuthNameFn({ oidcUser }),
  authRole: getAuthRoleFn({ oidcUser }),
  authUid: getAuthUidFn({ oidcUser }),
  isLoading: !oidcUser,
});

/**
 * UserProvider
 * @param {Function} useOidcUser
 * @param {JSX} children
 */

type TUserProvider = {
  children: ReactNode;
  useOidcUserFn: typeof useOidcUser;
  extractDataFromOAuthTokenFn?: typeof extractDataFromOAuthToken;
  isEnabled?: boolean;
};

const UserProvider = ({
  children,
  useOidcUserFn,
  extractDataFromOAuthTokenFn = extractDataFromOAuthToken,
  isEnabled = true,
  ...rest
}: TUserProvider) => {
  const { oidcUser } = useOidcUserFn();
  const value = useMemo(
    () => ({ ...extractDataFromOAuthTokenFn({ oidcUser }), isEnabled, ...rest }),
    [extractDataFromOAuthTokenFn, isEnabled, oidcUser, rest],
  );
  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default UserProvider;
