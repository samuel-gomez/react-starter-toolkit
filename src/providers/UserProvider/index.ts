export { default, UserContext } from './UserProvider';
export type { TUserContext } from './UserProvider';
