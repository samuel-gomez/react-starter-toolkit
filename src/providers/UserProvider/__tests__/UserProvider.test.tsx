import { useContext } from 'react';
import { render } from '@testing-library/react';
import omit from 'lodash/omit';
import UserProvider, { UserContext, getAuthName, getAuthRole, setAuthRole, getAuthUid } from '../UserProvider';

type TBase = {
  authName: string | undefined;
  authRole: string;
  authUid: string;
  isEnabled?: boolean;
  isLoading?: boolean;
};

const Base = ({ authName, authRole, authUid, isEnabled, isLoading }: TBase) => (
  <ul>
    <li>{isEnabled ? 'have isEnabled' : 'notHave isEnabled'}</li>
    <li>{isLoading ? 'have isLoading' : 'notHave isLoading'}</li>
    <li>{authName ? 'have authName' : 'notHave authName'}</li>
    <li>{authRole ? 'have authRole' : 'notHave authRole'}</li>
    <li>{authUid ? 'have authUid' : 'notHave authUid'}</li>
  </ul>
);

const BaseWithUser = () => {
  const userProps = useContext(UserContext);
  return <Base {...userProps} />;
};

const oidcUser = {
  name: 'Bob Smith',
  given_name: 'Bob',
  family_name: 'Smith',
  email: 'BobSmith@email.com',
  email_verified: true,
  website: 'http://bob.com',
  sub: '11',
  member_of: ['CN=Admin'],
  axa_uid_racf: 'S000007',
};

const profils = ['Admin', 'USER'];

const useOidcUserMock = jest.fn().mockReturnValue({
  oidcUser,
});

const App = () => (
  <UserProvider useOidcUserFn={useOidcUserMock}>
    <BaseWithUser />
  </UserProvider>
);

describe('Render App with Base have user props', () => {
  it('Should Base have user props when render App with UserProvider', () => {
    const { asFragment, getByText } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('have authName')).toBeDefined();
    expect(getByText('have authRole')).toBeDefined();
    expect(getByText('have authUid')).toBeDefined();
    expect(getByText('have isEnabled')).toBeDefined();
    expect(getByText('notHave isLoading')).toBeDefined();
  });
});

describe('getAuthName', () => {
  it('Should return "Bob Smith" when getAuthName called with profile name "Bob Smith" ', () => {
    const result = getAuthName({ oidcUser });
    expect(result).toEqual('Bob Smith');
  });

  it('Should return "" when getAuthName called with no name profile', () => {
    const result = getAuthName({ oidcUser: { ...omit(oidcUser, 'name') } });
    expect(result).toEqual('Non ConnectÃ©');
  });
});

describe('getAuthRole', () => {
  it('Should return Admin when getAuthRole called with profile member_of "CN=Admin" ', () => {
    const setAuthRoleFn = jest.fn().mockReturnValue('Admin');
    const result = getAuthRole({ oidcUser, setAuthRoleFn });

    expect(result).toEqual('Admin');
    expect(setAuthRoleFn).toHaveBeenCalledWith({ memberOf: oidcUser.member_of[0] });
  });

  it('Should return "" when getAuthRole called with no profile member_of', () => {
    const result = getAuthRole({ oidcUser: { ...omit(oidcUser, 'member_of') } });
    expect(result).toEqual('');
  });

  it('Should return "" when getAuthRole called with empty member_of', () => {
    const setAuthRoleFn = jest.fn().mockReturnValue('');
    const result = getAuthRole({ oidcUser: { ...oidcUser, member_of: [''] }, setAuthRoleFn });

    expect(result).toEqual('');
    expect(setAuthRoleFn).toHaveBeenCalledWith({ memberOf: '' });
  });
});

describe('getAuthUid', () => {
  it('Should return S000007 when getAuthUID called with profile axa_uid_racf = "S000007" ', () => {
    const result = getAuthUid({ oidcUser });
    expect(result).toEqual('S000007');
  });

  it('Should return "" when getAuthUID called with no profile axa_uid_racf', () => {
    const result = getAuthUid({ oidcUser: { ...omit(oidcUser, 'axa_uid_racf') } });
    expect(result).toEqual('');
  });
});

describe('setAuthRole', () => {
  it('Should return Admin when memberOf contain Admin ', () => {
    const result = setAuthRole({ memberOf: 'CN=Admin', profils });
    expect(result).toEqual('Admin');
  });

  it('Should return "" when memberOf not autorized', () => {
    const result = setAuthRole({ memberOf: 'CN=OTHER', profils });
    expect(result).toEqual('');
  });
});
