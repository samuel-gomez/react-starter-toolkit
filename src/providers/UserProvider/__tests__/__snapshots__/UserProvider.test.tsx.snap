// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render App with Base have user props Should Base have user props when render App with UserProvider 1`] = `
<DocumentFragment>
  <ul>
    <li>
      have isEnabled
    </li>
    <li>
      notHave isLoading
    </li>
    <li>
      have authName
    </li>
    <li>
      have authRole
    </li>
    <li>
      have authUid
    </li>
  </ul>
</DocumentFragment>
`;
